Extension/src/LanguageServer/copilotCompletionContextProvider.ts

```diff
@@ -0,0 +1,154 @@
+import { getCopilotApi } from "./copilotProviders";
+import { clients } from './extension';
+import { CodeSnippet, CompletionContext, ContextProviderApiV1 } from './tmp/contextProviderV1';
+
+// An ever growing cache of completion context snippets. //?? TODO Evict old entries.
+const completionContextCache: Map<string, CodeSnippet[]> = new Map<string, CodeSnippet[]>();
+const cppDocumentSelector: DocumentSelector = [{ language: 'cpp' }, { language: 'c' }];
+
+class DefaultValueFallback extends Error {
+    static readonly DefaultValue = "DefaultValue";
+    constructor() { super(DefaultValueFallback.DefaultValue); }
+}
+
+class CancellationError extends Error {
+    static readonly Cancelled = "Cancelled";
+    constructor() { super(CancellationError.Cancelled); }
+}
+
+let completionContextCancellation = new vscode.CancellationTokenSource();
+
+// Mutually exclusive values for the kind of snippets. They either come from the cache,
+// are computed, or the computation is taking too long and no cache is present. In the latter
+// case, the cache is computed anyway while unblocking the execution flow returning undefined.
+enum SnippetsKind {
+    Computed = 'computed',
+    CacheHit = 'cacheHit',
+    CacheMiss = 'cacheMiss'
+}
+
+// Get the default value if the timeout expires, but throws an exception if the token is cancelled.
+async function waitForCompletionWithTimeoutAndCancellation<T>(promise: Promise<T>, defaultValue: T | undefined,
+    timeout: number, token: vscode.CancellationToken): Promise<[T | undefined, SnippetsKind]> {
+    const defaultValuePromise = new Promise<T>((resolve, reject) => setTimeout(() => {
+        if (token.isCancellationRequested) {
+            reject('DefaultValuePromise was cancelled');
+        } else {
+            reject(new DefaultValueFallback());
+        }
+    }, timeout));
+    const cancellationPromise = new Promise<T>((_, reject) => {
+        token.onCancellationRequested(() => {
+            reject(new CancellationError());
+        });
+    });
+    let snippetsOrNothing: T | undefined;
+    try {
+        snippetsOrNothing = await Promise.race([promise, cancellationPromise, defaultValuePromise]);
+    } catch (e) {
+        if (e instanceof DefaultValueFallback) {
+            return [defaultValue, defaultValue !== undefined ? SnippetsKind.CacheHit : SnippetsKind.CacheMiss];
+        }
+
+        // Rethrow the error for cancellation cases.
+        throw e;
+    }
+
+    return [snippetsOrNothing, SnippetsKind.Computed];
+}
+
+// Get the completion context with a timeout and a cancellation token.
+// The cancellationToken indicates that the value should not be returned nor cached.
+async function getCompletionContextWithCancellation(documentUri: string, caretOffset: number,
+    startTime: number, out: Logger, token: vscode.CancellationToken): Promise<CodeSnippet[]> {
+    try {
+        const activeEditor: vscode.TextEditor | undefined = vscode.window.activeTextEditor;

```

Comment:
I don't think need you check the active text editor. You should just use the documentUri. The user could type and then switch editors, but the user/Copilot still wants the context for the original document and the current active document shouldn't matter.

Reply:
done, thanks!